### PR TITLE
ContextualMenu/ComboBox/Dropdown: Make the mouse being over a menu item not move focus until after a mouseMove is seen

### PR DESCRIPTION
**Pull request by jspurlin**
#### Pull request checklist

- [X] Include a change request file using `$ npm run change`

#### Description of changes
This PR addresses the fact that mouseEnter can fire when the menu is launched if the mouse is over a menu item. The fix is to not listen for mouseEnter/mouseLeave until a mouseMove has been detected first. (this is the case for CcntextualMenu, ComboBox, and Dropdown)

Before:
contextualmenumousemovebefore

After:
contextualmenumousemoveafter

Additionally, this also adds a prop to ContextualMenu to delay setting focus on any mouse event until the menu has focus. This fixes the case here a creator passes in false for shouldFocusOnMount. After this PR, if shouldFocusOnMount is false and delayUpdateFocusOnHover is true, the user can move the mouse over the menu without it updating focus until focus is set on the menu by some other means. Note, this is very similar to the first part of this change described, but without this fix the user experience is even worse since it moves focus into the menu when the creator did not expect it to be in the menu. The TellMe control in office is an example where this behavior can be seen.

#### Focus areas to test
Verified that ContextualMenu, ComboBox, and Dropdown no longer move focus when opened if the mouse is over a menu item, but once the mouse is moved focus is updated. Also verified that if delayUpdateFocusOnHover is not set (or is false) then focus is updated on mouseMove

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.base.tsx [hunk 1]
```diff
@@ -136,6 +140,13 @@ export class ContextualMenuBase extends BaseComponent<IContextualMenuProps, ICon
           : null;
       }
     }
+    if (newProps.delayUpdateFocusOnHover !== this.props.delayUpdateFocusOnHover) {
```
Comment (ddlbrena) [inline packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.base.tsx]: >if (newProps.delayUpdateFocusOnHover !== this.props.delayUpdateFocusOnHover) { [](start = 4, length = 78)

is the complexity introduced by delayUpdateFocusOnHover and _shouldUpdateFocusOnMouseEvent actually needed? Windows does not seem to do that and if the user moves the mouse at that point the user is mixing\switching input times so it that bad that focus moves? Seems that the complexity introduced here might not be worth the benefit?

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx [hunk 2]
```diff
@@ -1771,7 +1784,7 @@ export class ComboBox extends BaseComponent<IComboBoxProps, IComboBoxState> {
   };
 
   private _onOptionMouseEnter(index: number): void {
-    if (!this._isScrollIdle) {
+    if (!this._isScrollIdle || !this._gotMouseMove) {
```
Comment (ddlbrena) [inline packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: >(!this._isScrollIdle || !this._gotMouseMove) [](start = 7, length = 44)

perhaps move this logic into a function that you can share with _onOptionMouseLeave ? that way also you an name it which makes the code easier to read, something like shouldUpdatePendingValue or something

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx [hunk 3]
```diff
@@ -1781,6 +1794,10 @@ export class ComboBox extends BaseComponent<IComboBoxProps, IComboBoxState> {
   }
 
   private _onOptionMouseMove(index: number): void {
+    if (!this._gotMouseMove) {
```
Comment (ddlbrena) [inline packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: > if (!this._gotMouseMove) { [](start = 3, length = 27)

[nit] [nit] I think there is almost no perf difference that might as well just have this._gotMouseMove = true; here and get rid of the if check

File: packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.base.tsx (hunk 1 above)
Comment (jspurlin) [reply packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.base.tsx]: The TellMe control is exactly why this was added, it does a mix of keep focus in an input with the menu expanded and also moving focus from the input to the menu

Comment (jspurlin) [reply packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.base.tsx]: More generally, if someone creates a contextualMenu with shouldFocusOnMount = false, they may not want hover to update focus to the menu until they put focus in the menu

Comment (ddlbrena) [reply packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.base.tsx]: talked online, sounds good, I forgot about tellMe scenarios

---
In reply to: [213418687](https://github.com/OfficeDev/office-ui-fabric-react/pull/6109#discussion_r213418687) [](ancestors = 213418687)

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx (hunk 2 above)
Comment (jspurlin) [reply packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: pulled it into a shouldIgnoreMouseEvent function

**Comment by jspurlin** [comment]
@betrue-final-final can you approve the screener update? Note, with this change a menu item doesn't expand on hover until a mousemove happens (hence the difference in the screener screenshot)

**Comment by betrue-final-final** [comment]
Nevermind. I got it from PR details.

**Pull request merged**